compiler/rustc_hir_analysis/src/collect.rs

```diff
@@ -346,6 +350,94 @@ impl<'tcx> HirTyLowerer<'tcx> for ItemCtxt<'tcx> {
         ty::Const::new_error_with_message(self.tcx(), span, "bad placeholder constant")
     }
 
+    fn lower_field_path(
+        &self,
+        container: &rustc_hir::Ty<'tcx>,
+        fields: &[Ident],
+        span: Span,
+        hir_id: HirId,
+        field_path_kind: FieldPathKind,
+    ) -> Result<(Ty<'tcx>, FieldPath<'tcx>), ErrorGuaranteed> {
+        assert_eq!(field_path_kind, FieldPathKind::FieldOf);
+        let container = self.lower_ty(container);
+
+        let mut field_indices = Vec::with_capacity(fields.len());
+        let mut current_container = container;
+        let mut fields = fields.into_iter().peekable();
+        let mut infcx = None;
+        let infcx = self.infcx().unwrap_or_else(|| {
+            assert!(!container.has_infer());
+            infcx = Some(self.tcx().infer_ctxt().build(ty::TypingMode::non_body_analysis()));
+            infcx.as_ref().unwrap()
+        });
+        let cause = ObligationCause::misc(span, self.item_def_id);
+        let at = infcx.at(&cause, ParamEnv::empty());
+        let ty_ch_cx = RefCell::new(<dyn TraitEngine<'_, FulfillmentError<'tcx>>>::new(&infcx));
+
+        while let Some(&field) = fields.next() {
+            let last = fields.peek().is_none();
+            let result =
+                at.structurally_normalize_ty(current_container, &mut **ty_ch_cx.borrow_mut());
```

Comment:
I am curious @lcnr. Do you think we can normalise types with erased lifetimes here, assuming that we would not need to check any lifetime bounds and we would only concern about the shapes of ADT?